### ISSUE TITLE
Missing POA middleware

### ISSUE THREAD

**Issue by cygnusv** [nucypher/nucypher#3525]
**Describe the Bug**
Several nodes running v7.3.0 have described getting the missing POA middleware error from web3py. This has blocked them from making TXs to the network.

Is it possible that this is a ATXM error? I thought we fixed this error in NuCypher a long time ago. Also, is it possible that this is some sort of sporadic error from a misconfiguration at the RPC provider?

**To Reproduce**
We don't have steps to reproduce for the moment

**Traceback or Screenshots (Optional)**
From a node operator:
```
    web3.exceptions.ExtraDataLengthError: The field extraData is 281 bytes, but should be 32. It is quite likely that you are connected to a POA chain. Refer to http://web3py.readthedocs.io/en/stable/middleware.html#proof-of-authority for more details. The full extraData is:
SKIPPED

2024-07-23T15:01:05+0000 [AutomaticTxMachine#warn] [recovery] restarting transaction machine!
2024-07-23T15:01:05+0000 [AutomaticTxMachine#info] [atxm] starting async transaction machine in 300 seconds
```
From another node operator:
```
 web3.exceptions.ExtraDataLengthError: The field extraData is 165 bytes, but should be 32. It is quite likely that you are connected to a POA chain. Refer to http://web3py.readthedocs.io/en/stable/middleware.html#proof-of-authority for more details. The full extraData is: HexBytes('0xd78301030383626f7288676f312e32322e33856c696e75780000000000000000f84280f83fc0c0c180c0c0c0c0c106c0c104c0c10ac0c102c10dc10ec0c105c0c10fc113c0c114c0c116c0c0c11ac11bc11cc0c118c0c21119c120c11fc123c121c125c07423611e3f05e9fb83f89bbe2e9e98ec2c974caab7797084354e1d674e1a1df00722fcba26c009ce3d05f8d8afa8b1e236ac8d6fa11fe4428e62165eb9ab29ec00')
 ```

**System (please complete the following information):**
 - OS Version:
 - Nucypher Version:
 - Ethereum Node Version:

**Comment by KPrasch** [comment]
> Is it possible that this is a ATXM error? 

I'm leaning towards this being an atxm error since it happens under the atxm stack seem below.

Here's a full traceback: <https://gist.github.com/KPrasch/c8d8b9972b664f11ef17d17602193067>

**Comment by cygnusv** [comment]
It happened again. Node running v7.4.0 and using Infura.

```
2024-08-27T16:57:22+0000 [twisted.python.log#info] "188.166.27.46" - - [27/Aug/2024:16:57:21 +0000] "POST /node_metadata HTTP/1.1" 200 82 "-" "python-requests/2.32.3"
2024-08-27T16:57:24+0000 [operator#info] Learning round 149506.  Teacher: (Ursula)⇀Orange Romeo SeaShell Hotel↽ (0xc8A6Bfe0834FB99340a0Df253d79B7CaE25053b8) knew about 128 nodes, 0 were new.
2024-08-27T16:57:27+0000 [twisted.python.log#info] "209.97.147.51" - - [27/Aug/2024:16:57:26 +0000] "POST /node_metadata HTTP/1.1" 200 82 "-" "python-requests/2.32.3"
2024-08-27T16:57:29+0000 [operator#info] Learning round 149507.  Teacher: (Ursula)⇀SlateBlue Whiskey MediumBlue Kilo↽ (0xDC09db6e5DA859eDeb7FC7bDCf47545056dC35F7) knew about 128 nodes, 0 were new.
2024-08-27T16:57:36+0000 [twisted.python.log#info] "164.92.141.186" - - [27/Aug/2024:16:57:35 +0000] "POST /node_metadata HTTP/1.1" 200 82 "-" "python-requests/2.31.0"
2024-08-27T16:57:37+0000 [AutomaticTxMachine#warn] [recovery] error during transaction: Traceback (most recent call last):
	  File "/usr/local/lib/python3.12/site-packages/twisted/internet/base.py", line 695, in run
	    self.mainLoop()
	  File "/usr/local/lib/python3.12/site-packages/twisted/internet/base.py", line 705, in mainLoop
	    self.runUntilCurrent()
	  File "/usr/local/lib/python3.12/site-packages/twisted/internet/base.py", line 1090, in runUntilCurrent
	    call.func(*call.args, **call.kw)
	  File "/usr/local/lib/python3.12/site-packages/twisted/internet/task.py", line 251, in __call__
	    d = maybeDeferred(self.f, *self.a, **self.kw)
	--- <exception caught here> ---
	  File "/usr/local/lib/python3.12/site-packages/twisted/internet/defer.py", line 212, in maybeDeferred
	    result = f(*args, **kwargs)
	  File "/usr/local/lib/python3.12/site-packages/atxm/machine.py", line 210, in _cycle
	    self._cycle_state()
	  File "/usr/local/lib/python3.12/site-packages/statemachine/event.py", line 34, in trigger_event
	    result = event_instance.trigger(self, *args, **kwargs)
	  File "/usr/local/lib/python3.12/site-packages/statemachine/event.py", line 27, in trigger
	    return machine._processing_loop()
	  File "/usr/local/lib/python3.12/site-packages/statemachine/statemachine.py", line 123, in _processing_loop
	    return self._engine.processing_loop()
	  File "/usr/local/lib/python3.12/site-packages/statemachine/engines/sync.py", line 72, in processing_loop
	    result = self._trigger(trigger_data)
	  File "/usr/local/lib/python3.12/site-packages/statemachine/engines/sync.py", line 104, in _trigger
	    result = self._activate(event_data)
	  File "/usr/local/lib/python3.12/site-packages/statemachine/engines/sync.py", line 120, in _activate
	    result = self.sm._get_callbacks(transition.before.key).call(*args, **kwargs)
	  File "/usr/local/lib/python3.12/site-packages/statemachine/callbacks.py", line 327, in call
	    callback.call(*args, **kwargs)
	  File "/usr/local/lib/python3.12/site-packages/statemachine/callbacks.py", line 273, in call
	    value = self._callback(*args, **kwargs)
	  File "/usr/local/lib/python3.12/site-packages/statemachine/signature.py", line 65, in signature_adapter
	    return method(*ba.args, **ba.kwargs)
	  File "/usr/local/lib/python3.12/site-packages/atxm/machine.py", line 180, in _enter_busy_mode
	    average_block_time = _get_average_blocktime(
	  File "/usr/local/lib/python3.12/site-packages/atxm/utils.py", line 22, in _get_average_blocktime
	    latest_block = w3.eth.get_block("latest")
	  File "/usr/local/lib/python3.12/site-packages/web3/eth/eth.py", line 413, in get_block
	    return self._get_block(block_identifier, full_transactions)
	  File "/usr/local/lib/python3.12/site-packages/web3/module.py", line 75, in caller
	    result = w3.manager.request_blocking(
	  File "/usr/local/lib/python3.12/site-packages/web3/manager.py", line 329, in request_blocking
	    response = self._make_request(method, params)
	  File "/usr/local/lib/python3.12/site-packages/web3/manager.py", line 213, in _make_request
	    return request_func(method, params)
	  File "/usr/local/lib/python3.12/site-packages/nucypher/blockchain/middleware/retry.py", line 75, in __call__
	    raise result
	  File "/usr/local/lib/python3.12/site-packages/nucypher/blockchain/middleware/retry.py", line 52, in __call__
	    response = self.make_request(method, params)
	  File "/usr/local/lib/python3.12/site-packages/web3/middleware/gas_price_strategy.py", line 101, in middleware
	    return make_request(method, params)
	  File "/usr/local/lib/python3.12/site-packages/web3/middleware/formatting.py", line 126, in middleware
	    response = make_request(method, params)
	  File "/usr/local/lib/python3.12/site-packages/web3/middleware/attrdict.py", line 43, in middleware
	    response = make_request(method, params)
	  File "/usr/local/lib/python3.12/site-packages/web3/middleware/formatting.py", line 128, in middleware
	    return _apply_response_formatters(
	  File "cytoolz/functoolz.pyx", line 263, in cytoolz.functoolz.curry.__call__
	    return self.func(*args, **kwargs)
	  File "/usr/local/lib/python3.12/site-packages/web3/middleware/formatting.py", line 74, in _apply_response_formatters
	    return _format_response("result", result_formatters[method])
	  File "/usr/local/lib/python3.12/site-packages/web3/middleware/formatting.py", line 70, in _format_response
	    response, response_type, method_response_formatter(appropriate_response)
	  File "cytoolz/functoolz.pyx", line 263, in cytoolz.functoolz.curry.__call__
	    return self.func(*args, **kwargs)
	  File "/usr/local/lib/python3.12/site-packages/eth_utils/applicators.py", line 72, in apply_formatter_if
	    return formatter(value)
	  File "cytoolz/functoolz.pyx", line 263, in cytoolz.functoolz.curry.__call__
	    return self.func(*args, **kwargs)
	  File "/usr/local/lib/python3.12/site-packages/eth_utils/functional.py", line 45, in inner
	    return callback(fn(*args, **kwargs))
	  File "/usr/local/lib/python3.12/site-packages/eth_utils/applicators.py", line 84, in apply_formatters_to_dict
	    yield key, formatters[key](item)
	  File "/usr/local/lib/python3.12/site-packages/web3/middleware/validation.py", line 76, in _check_extradata_length
	    raise ExtraDataLengthError(
	web3.exceptions.ExtraDataLengthError: The field extraData is 1158 bytes, but should be 32. It is quite likely that you are connected to a POA chain. Refer to http://web3py.readthedocs.io/en/stable/middleware.html#proof-of-authority for more details. The full extraData is: HexBytes('0xd78301030783626f7288676f312e32322e35856c696e75780000000000000000f90422b903980308e8b459e3f91f25b855a3d429af2883c505c40000000000000000000000000000000000000001048cfedf907c4c9ddd11ff882380906e78e84bbe0000000000000000000000000000000000000002127685d6dd6683085da4b6a041efcef1681e5c9c0000000000000000000000000000000000000002144aeae662aeb1f871ac217bb310d241c0f5059000000000000000000000000000000000000000021b0840519a581f3779d0a10b77593d6d3894a76a00000000000000000000000000000000000000011d25c827abd466387bda00b429fe728627d6eee600000000000000000000000000000000000000021efecb61a2f80aa34d3b9218b564a64d0594629000000000000000000000000000000000000000024b1b3e74d843bc6b3f3648c7957252104bda3f1a00000000000000000000000000000000000000015627fe906f34a53a381a726dbbfff03f65ebcfb5000000000000000000000000000000000000000167b94473d81d0cd00849d563c94d0432ac988b4900000000000000000000000000000000000000097c7379531b2aee82e4ca06d4175d13b9cbeafd49000000000000000000000000000000000000000183d69448f88bf9c701c1b93f43e1f753d39b263200000000000000000000000000000000000000029ead03f7136fc6b4bdb0780b00a1c14ae5a8b6d00000000000000000000000000000000000000002a8b52f02108aa5f4b675bdcc973760022d7c60200000000000000000000000000000000000000004aa6ac02fddaaf6f120f5bb98ce30809d19cd5d1b0000000000000000000000000000000000000002b2dd091ea6e591d62f565d7a18ce2a7640add2270000000000000000000000000000000000000001b95d435df3f8b2a8d8b9c2b7c8766c9ae6ed8cc90000000000000000000000000000000000000002b9ede6f94d192073d8eaf85f8db677133d4832490000000000000000000000000000000000000001c35649ae99be820c7b200a0add09b96d7032d2320000000000000000000000000000000000000001c84c1e8dffe81c8f69f2b655557ef57d5b46beb50000000000000000000000000000000000000001ec20607aa654d823dd01beb8780a44863c57ed070000000000000000000000000000000000000001eedba2484aaf940f37cd3cd21a5d7c4a7dafbfc00000000000000000000000000000000000000006f0245f6251bef9447a08766b9da2b07b28ad80b00000000000000000000000000000000000000003f885c0c0c101c0c101c104c104c106c0c0c0c108c0c2070cc0c10dc10fc0c110c110c0c0c112c0c21316c118c119c0c117c0c11ac0c11ec120c0c0c0c2010bc0c121c0c127c129c0c0c125c125c0c11cc0c105c22532c2332ec0c12ac3303428c23719c138c23805c0c0c138c23d36c23a3ec0c13fc2383ec22d3fc132c24443c131c22d42c0c0fc2b78a63b27a75f4553c124aeaa499a6b52e8a7938cbf2c49a3e4943b7d8378435b24ad454d00c220262810b8ca16c79492e75cc76bff4610263075a7a1cbe501')
	
2024-08-27T16:57:37+0000 [AutomaticTxMachine#warn] [recovery] restarting transaction machine!
2024-08-27T16:57:37+0000 [AutomaticTxMachine#info] [atxm] starting async transaction machine in 300 seconds

````

**Comment by cygnusv** [comment]
I've observed that simply restarting the node can help in these situations. This makes me thing that:
* Is it possible there's some type of race condition happening that's preventing from getting the middlewares correctly in the first place?
* A potential countermeasure is to automatically restart the node when this error is detected

**Comment by derekpierre** [comment]
Just adding one more point:

* There is an element of time here too. In many of the cases, we observe the error after the nodes have been running for a long time i.e. nodes don't ever call `eth.get_block(...)` (for which the poa middleware is needed) until they are sampled for a ritual. So it could be days, weeks of running after start up before an `eth.get_block()`  call is made. When nodes are part of a ritual ceremony and get restarted (due to an error) during the ceremony, they call `eth.get_block` right after restart and this call works. I'm wondering if the poa middleware is correctly added, but perhaps the time taken to call `eth.get_block` and have the poa be used is playing a role..?

**Comment by derekpierre** [comment]
Closed via #3549
